### DISCUSSION TITLE
Unable to make connection to the Websocket for transcribing live stream audio

### DISCUSSION THREAD

**Discussion by SttlUser** [deepgram/community#506] [General help] [answered]
Hi Team, 
I am trying to transcribe the audio via live stream audio transcription endpoint "wss://api.deepgram.com/v1/listen". As per the documentation I need to first establish a connection with the websocket before being able to transcribe the audio, I have tried that but i am constantly getting the same error as below:

> in handshake
    raise InvalidStatusCode(status_code, response_headers)
websockets.exceptions.InvalidStatusCode: server rejected WebSocket connection: HTTP 401

I have provide the python code snippet below for reference:

  

> import asyncio
>   import websockets
>   import json
>   async def send_audio(api_key):
>       async with websockets.connect("wss://api.deepgram.com/v1/listen?model=nova-2-general&language=en-US", extra_headers=[("Authorization", f"API_KEY {api_key}")]) as websocket:
>           await websocket.send(json.dumps({"type": "KeepAlive"}))
>           while True:
>               await asyncio.sleep(7) 
>               await websocket.send(json.dumps({"type": "KeepAlive"}))
>   your_api_key = "****************"
>   asyncio.get_event_loop().run_until_complete(send_audio(your_api_key))

Request your assistance...

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Reply by nikolawhallon** [reply]
My first reaction is to double check the api key, the HTTP 401 error means the request is unauthorized (bad api key).

**Reply by SttlUser** [reply]
Well I did that and to be sure I have created a new API with owner permissions. but I am still setting the same error just with the 400 error.

**Reply by nikolawhallon** [reply]
Are you now getting an HTTP 400 error or still an HTTP 401 error? Can you share the full output/error?

**Reply by SttlUser** [reply]
Well, I was able to resolve the 401 error but I am still facing some issue. I was trying to fetch the API_KEY from ENV file but it always passed the empty value. anyways, after resolving this issue. I was able to connect to the WebSocket but when ever I try to transcribe the audio, I am getting empty response. for example, 

> Transcription: {'type': 'Results', 'channel_index': [0, 1], 'duration': 0.52000046, 'start': 9.62, 'is_final': True, 'speech_final': True, 'channel': {'alternatives': [{'transcript': '', 'confidence': 0.0, 'words': []}]}, 'metadata': {'request_id': '**************************', 'model_info': {'name': '2-general-nova', 'version': '2023-11-14.6411', 'arch': 'nova-2'}, 'model_uuid': '***********************'}}

**Reply by nikolawhallon** [reply]
Can you share the full output and a request id for the problematic streams? Not every message will contain transcription results (since during the stream probably there are times when no words are being spoken), but if you never receive any transcription over the whole stream, I would imagine the audio being sent has issues - we have a streaming test suite which can help debug audio issues, but given a request id I can look on our end to see if I can find the issue (https://github.com/deepgram/streaming-test-suite).

**Reply by SttlUser** [reply]
Thank @nikolawhallon for the response, I can surely provide you the request Id but I have tested the system microphone, it is working properly if that's what you are trying to say. Also I do not get any transcription for the whole stream, I have not tried the test suit but now I think I will go through it.

**Reply by davidvonthenen** [reply]
If you still need help with this, please let me know.

**Reply by SttlUser** [reply]
Well @dvonthenen, I am still looking for some help but not on this topic any more. I am using the library instead of API endpoint and it is working properly but I need some help configuring it. should I create a new thread or should I just continue on this?

**Reply by davidvonthenen** [reply]
if you want real-time-ish help, stop by discord. if you want to keep it here, just continue to use this same thread

**Reply by SttlUser** [reply]
Well I did try to connect over discord but I could not join in...

**Reply by SttlUser** [reply]
I am trying to create a python application similar to a chatbot. I just want the deepgram model to listen till the silence is detected. I assume there will be some parameter or a way to do it. I have taken reference from deepgram website but the reference code script was running for defined time only(so the model will only listen till 5s, 10s, ... only).

**Reply by nikolawhallon** [reply, marked as answer]
Check out these docs for info on detecting the end of speech during streaming mode: https://developers.deepgram.com/docs/understanding-end-of-speech-detection-while-streaming#using-utteranceend-and-endpointing

note that in streaming mode, users can always start speaking again, even after they take a pause, but that doc goes over some of the caveats and nuances there
